torch/_inductor/ir.py

```diff
@@ -5297,6 +5297,12 @@ def __init__(
             self.mutation_names.append(tensor_args[0].get_name())
             return
 
+        if self.op_overload is torch.ops.fsdp.split_with_sizes_copy.default:
```

Comment:
How does this node get inserted in the first place ?

I was going to say something along the lines of generalize this mutation analysis but it doesn't seem like this will occur very often:

```
# The unsupported cases usually do not show up here (because
# AOTAutograd functionalized them away); the only way for an in-place
# op to show up here is if a lowering or pass introduced it.```

Reply:
yes we intentionally preserve this op through the lowering stack (no functionalization, no decomposition): https://github.com/pytorch/pytorch/pull/129414/files: no functionalization to more easily work around AOTAutograd limitations, no decomp so that we can do pattern-matching on it much more easily (in future PR).

would love to make this more general - maybe I should create a dictionary where each op maps to its specific mutation analysis? (although there will only be one op `fsdp.split_with_sizes_copy` for now)

Reply:
I'm confused about the strategy here. How do we ensure there is no silent incorrectness if we preserve this operator through the PT2 stack without functionalization? I would have expected that we auto-functionalize it and then later re-inplace it.

Reply:
Let me give some background first: Will was running into an issue in fsdp where he had a graph (from compiled autograd) with the following traits:

(1) two inputs of the graph were aliased to each other
(2) one of them is mutated (a resize_ and set_ to handle the all-gathered param)
(3) they are both subclasses

This situation is "handled" in AOTAutograd today, but (a) not in a very performant way (we emit `as_strided` and `as_strided_scatter` in several places) and (b) we error on subclasses, so this doesn't work at all for FSDP.

The compromise we were looking into was "de-functionalization", for a small number of mutable ops that only FSDP cares about: `resize_` and `set_`. This "solves" the problem above, because from AOTAutograd pov there are no mutations that functionalization needs to handle. "Not functionalizing" also has some requirements though:

(1) we need to not DCE the mutable op
(2) inductor needs to handle this mutable op directly in its IR (previously, mutable ops could only show up as `copy_` or `set_` nodes, and only showed up in a small epilogue at the end of the graph)

I think we handle (1) properly today, just by adding the two specific ops to FX's allow-list. I'm not sure what the status of (2) is (although this change is probably trying to improve it).

For `fsdp.split_with_sizes_copy.out`: this is the 3rd "interesting" mutable op that FSDP cares about. @yf225 will remember better, but if this op gets invoked on one of the aliased inputs, then I think we'll need to de-functionalize it to avoid the AOT issue mentioned above. If this op is not actually invoked on either of the aliased inputs, then in theory we could do the normal auto-functionalize flow for custom ops on it. Although the reinplacing pass will be load-bearing for FSDP, so we probably will want some tests that reinplacing always works properly on that op.

@zou3519 let me know what you think

Reply:
my other baked-in assumption: if we are already de-functionalizing `set_` and `resize_` that already requires a lift from inductor to handle mutable ops in the middle of the graph, so "one extra mutable op" isn't too much of a lift. Does that assumption sound wrong?

Reply:
My take is that morally, "de-functionalizing" should be a last resort (otherwise, why not de-functionalize *all* custom ops?). Since `fsdp.split_with_sizes_copy.out` looks like a standard op from the outside, we should functionalize it via our standard mechanism (auto_functionalize).

I expected "de-functionalizing" to cause some sort of barrier mechanism in inductor that would be harmful to performance, hence why it should be a last resort.
(for (2) -- naively I would have expected that we partition the graph based on the set_, pass each functional subgraph to inductor's post_grad passes, and then stitch the graph back together, which means there is a performance implication for not being able to optimize across `set_` calls).

Side note: I want to rename "de-functionalizing" -- it's more like "forgoing functionalization"

Reply:
> For fsdp.split_with_sizes_copy.out: this is the 3rd "interesting" mutable op that FSDP cares about. @yf225 will remember better, but if this op gets invoked on one of the aliased inputs, then I think we'll need to de-functionalize it to avoid the AOT issue mentioned above

I don't fully understand the problem. Is the claim that auto_functionalization will do something subpar, so we need to not do it?

Reply:
Discussed with @zou3519 offline, the details are:

It might look like `fsdp.split_with_sizes_copy` is a standard op so auto-functionalize should just work. But the reality is that `fsdp.split_with_sizes_copy` is (transitively) mutating a graph input but auto-functionalize doesn't know about it, and hence it will not issue the `.copy_` to copy the output back to the original tensor to preserve semantics:
```python
def f(x, y):
    tmp = torch.empty_like(out)
    res = auto_functionalized(op, x, out=tmp)
    # we need a tmp.copy_(res) here, but we don't have it
    y.set_(tmp)
```

Reply:
Updated the PR to go with the original direction: not functionalizing `fsdp.split_with_sizes_copy`.

Reply:
I repro'd the silent correctness locally in that test when running with `aot_eager`: I think we can keep using auto-functionalize for `split_with_sizes_copy`, but there was a bug in the "do-not-functionalize" logic for set_. This patch fixed the test for me:
```
diff --git a/torch/distributed/_composable/fsdp/_fsdp_param.py b/torch/distributed/_composable/fsdp/_fsdp_param.py
index 955beb86699..866108030d0 100644
--- a/torch/distributed/_composable/fsdp/_fsdp_param.py
+++ b/torch/distributed/_composable/fsdp/_fsdp_param.py
@@ -76,6 +76,8 @@ def set_(tensor, data):

 @torch.library.impl(lib, "set_", "Functionalize")
 def set__functionalize(tensor, data):
+    torch._sync(tensor)
+    torch._sync(data)
     tensor_inner = torch._from_functional_tensor(tensor)
     data_inner = torch._from_functional_tensor(data)
     tensor_inner.set_(data_inner)  # type: ignore[call-overload]
```

I ended up making a minimal repro below that shows the same correctness problem. This issue is that when we dump `set_` into the otherwise-functional graph, we need to make sure that its inputs have had any pending mutations properly "flushed". Otherwise, the inputs that we pass into the `set_` will be the values of those inputs before any mutations were performed on them.
```
import torch
import numpy as np

import torch.distributed._composable.fsdp._fsdp_param

@torch.library.custom_op("mylib::add_one_out", mutates_args={"out"})
def add_one_out(x: torch.Tensor, out: torch.Tensor) -> None:
    torch.add(x, 1, out=out)

@torch.compile(backend="aot_eager")
def f(x):
    buf = torch.zeros(2)
    buf_view = buf.view(-1)
    torch.ops.mylib.add_one_out(x, out=buf_view)
    buf_view2 = buf.view(-1)
    torch.ops.fsdp.set_(x, buf_view2)

x = torch.zeros(2)
f(x)
print(x)
```

Reply:
As Brian mentioned, fixing the `set_` "do-not-functionalize" logic allows us to use auto-functionalize for `split_with_sizes_copy`. We took this approach in https://github.com/pytorch/pytorch/pull/129502 and were unblocked on enabling E2E inductor unit test.